File: swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/Telemetry.swift
//
//  Telemetry.swift
//  (c) 2024 Firezone, Inc.
//  LICENSE: Apache-2.0
//

import Sentry

public enum Telemetry {
  /// Sets the Sentry user from the firezone device ID and account slug.
  public static func setUser(firezoneId: String, accountSlug: String) {
    SentrySDK.configureScope { scope in
      let user = User(userId: firezoneId)
      user.data = ["account_slug": accountSlug]
      scope.setUser(user)
    }
  }

  public static func start(enableAppHangTracking: Bool = true) {
    SentrySDK.start { options in
      options.dsn =
        "https://66c71f83675f01abfffa8eb977bcbbf7@o4507971108339712.ingest.us.sentry.io/4508175177023488"
      options.environment = "entrypoint"  // will be reconfigured in VPNConfigurationManager
      options.releaseName = releaseName()
      options.dist = distributionType()
      options.enableAppHangTracking = enableAppHangTracking
    }
  }

  public static func setEnvironmentOrClose(_ apiURL: String) {
    var environment: String?

    if apiURL.starts(with: "wss://api.firezone.dev") {
      environment = "production"
    } else if apiURL.starts(with: "wss://api.firez.one") {
      environment = "staging"
    }

    guard let environment
    else {
      // Disable Sentry in unknown environments
      SentrySDK.close()

      return
    }

    SentrySDK.configureScope { configuration in
      configuration.setEnvironment(environment)
    }
  }

  public static func capture(_ err: Error) {
    SentrySDK.capture(error: err)
  }

  private static func distributionType() -> String {
    // Apps from the app store have a receipt file
    if BundleHelper.isAppStore() {
      return "appstore"
    }

    return "standalone"
  }

  private static func releaseName() -> String {
    let version =
      Bundle.main.infoDictionary?["CFBundleShortVersionString"]
      as? String ?? "unknown"

    #if os(iOS)
      return "ios-client@\(version)"
    #else
      return "macos-client@\(version)"
    #endif
  }
}

File: kotlin/android/app/src/main/java/dev/firezone/android/core/Telemetry.kt
// Licensed under Apache 2.0 (C) 2026 Firezone, Inc.
package dev.firezone.android.core

import android.content.Context
import android.os.Build
import dev.firezone.android.BuildConfig
import io.sentry.Sentry
import io.sentry.android.core.SentryAndroid
import io.sentry.protocol.User

object Telemetry {
    private var firezoneId: String? = null
    private var accountSlug: String? = null
    private var accountId: String? = null
    private var actorEmail: String? = null
    private var mdmDeviceId: String? = null

    fun start(context: Context) {
        SentryAndroid.init(context) { options ->
            options.dsn =
                "https://928a6ee1f6af9734100b8bc89b2dc87d@sentry.firezone.dev/4508175126233088"
            options.environment = "entrypoint"
            options.release = releaseName()
            options.dist = distributionType(context)
            options.logs.isEnabled = true
        }
    }

    fun setUser(
        firezoneId: String,
        accountSlug: String,
    ) {
        this.firezoneId = firezoneId
        this.accountSlug = accountSlug
        accountId = null
        actorEmail = null
        updateUser()
    }

    fun setUser(
        firezoneId: String,
        accountId: String,
        actorEmail: String,
    ) {
        this.firezoneId = firezoneId
        accountSlug = null
        this.accountId = accountId
        this.actorEmail = actorEmail
        updateUser()
    }

    fun setMdmDeviceId(id: String?) {
        mdmDeviceId = id
        updateUser()
    }

    private fun updateUser() {
        val id = firezoneId
        val slug = accountSlug
        val accountId = accountId
        val actorEmail = actorEmail
        val mdmId = mdmDeviceId

        if (id != null && (slug != null || (accountId != null && actorEmail != null))) {
            Log.setUser(id, slug, accountId, actorEmail, mdmId)
            val user =
                User().apply {
                    this.id = id
                    this.email = actorEmail
                    this.data =
                        buildMap {
                            slug?.let { put("account_slug", it) }
                            accountId?.let { put("account_id", it) }
                            actorEmail?.let { put("actor_email", it) }
                            mdmId?.let { put("mdm_device_id", it) }
                        }
                }
            Sentry.setUser(user)
        } else {
            Log.clearUser()
            Sentry.setUser(null)
        }
    }

    fun setEnvironmentOrClose(apiUrl: String) {
        val environment =
            when {
                apiUrl.startsWith("wss://api.firezone.dev") -> "production"
                apiUrl.startsWith("wss://api.firez.one") -> "staging"
                else -> null
            }

        if (environment != null) {
            Log.setEnvironment(environment)
            Sentry.getCurrentScopes().options.environment = environment
        } else {
            Sentry.close()
        }
    }

    fun capture(error: Throwable) {
        Sentry.captureException(error)
    }

    private fun distributionType(context: Context): String =
        try {
            val installer =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    context.packageManager.getInstallSourceInfo(context.packageName).installingPackageName
                } else {
                    @Suppress("DEPRECATION")
                    context.packageManager.getInstallerPackageName(context.packageName)
                }
            if (installer == "com.android.vending") "google_play" else "standalone"
        } catch (_: Exception) {
            "standalone"
        }

    private fun releaseName(): String {
        val version = BuildConfig.VERSION_NAME
        return "android-client@$version"
    }
}
